### ISSUE TITLE
Discussion: Difficulty adjustment does not factor in composing time.  unclear if a problem or not.

### ISSUE THREAD

**Issue by dan-da** [Neptune-Crypto/neptune-core#263]
The difficulty_control() fn that calculates new difficulty for each block has signature:

```
pub(crate) fn difficulty_control(
    new_timestamp: Timestamp,
    old_timestamp: Timestamp,
    mut old_difficulty: Difficulty,
    target_block_interval: Option<Timestamp>,
    previous_block_height: BlockHeight,
) -> Difficulty
```

It is unaware of how much time was spent composing for each block.   So let's say that target_block_interval is 10 minutes, but composing takes 11 minutes.   In this case, the difficulty will never go up.   maybe that's ok/correct.

Probably the more common case is that composing takes say 4 minutes and target is 10 minutes.  Perhaps it is still ok, because difficulty continues to adjust each block and is able to manipulate the guessing to take approx 6 minutes.   ie, the difficulty is perpetually lower than it would be if guessing took 100% of the time

One possible issue I see is if the composing time is variable between blocks.  The difficulty adjustment has no knowledge of time spent composing vs guessing, ie the ratio.  It is calculating the new difficulty assuming 100% of time is spent guessing, but really it may be 50% one block, 30% another block, 65% another block, and so on.  So maybe we will have higher block-time variability than we would at 100% guessing.

I'm unsure if this is going to create any problem or it just kinda "works out" ok.

Opening this issue for comment/discussion.

**Comment by dan-da** [comment]
I'm presently testing a multi-threaded mining (guessing) patch.  Running a node that is both a composer and a guesser, starting from genesis block.   The blockchain has reached height 14.  So far every block has been found instantly, within milliseconds once guessing begins.  The difficulty is adjusting upwards, but I think it is doing so much more slowly than it used to when guessing took ~100% of the time.    Composing has been taking between 300 and 500 secs, ie 5 to 8.3 minutes.

**Comment by aszepieniec** [comment]
It's probably a good idea to investigate this further. I just want to add some context.

>  The difficulty adjustment has no knowledge of time spent composing vs guessing, ie the ratio. It is calculating the new difficulty assuming 100% of time is spent guessing

The difficulty adjustment does not make any assumptions about the ratio between the time spent composing and the time spent guessing. And so concretely: if the same sequence of timestamps had occurred, then the difficulty adjustments would have been identical -- regardless of whether 100%, 50%, or 1% of the block times had been spent guessing.

> really it may be 50% one block, 30% another block, 65% another block, and so on. So maybe we will have higher block-time variability than we would at 100% guessing.

I don't know where the variance in composing time comes from. Regardless, it is safe to assume that ratio of composing time will tend towards something insignificant with better algorithmic and hardware support for composing. In this steady state, almost 100% of the variance will come from the guessing step.

> The difficulty is adjusting upwards, but I think it is doing so much more slowly than it used to when guessing took ~100% of the time.

Did you come to this conclusion by comparing a recent experience to one several weeks in the past, or by comparing two recent experiences whereby the composing time was artificially set to zero? I ask because a couple of weeks ago the scaling factor was changed and a minimum block time was enforced; these changes could explain the former observation.

It also worth noting that the magnitude of the adjustment is larger if two blocks are found in closer succession. So with blocks found almost instantly, you would expect the difficulty to rise very fast.

**Comment by Sword-Smith** [comment]
It takes some time for difficulty to adjust fully, maybe around 100 blocks? I don't think there is an issue here.

**Comment by dan-da** [comment]
> Did you come to this conclusion by comparing a recent experience to one several weeks in the past

yes, before proving landed.   At that time, guessing took up ~100% of the time between blocks.  And when mining regtest from genesis block, the first block would be mined instantly but by height 2 it was taking a minute or more to mine.    So the difficulty adjustment was kicking in quickly.     For comparison, last night I mined 58 blocks, and the guessing portion was still just a few milliseconds.    The node was killed overnight so it didn't run long enough to observe any  obvious lengthening of the block times.

> It also worth noting that the magnitude of the adjustment is larger if two blocks are found in closer succession. So with blocks found almost instantly, you would expect the difficulty to rise very fast.

Right, but on this machine composing takes 5-10 minutes, so blocks are never found instantly.  Only the guessing portion is almost instant.  So I believe that makes the adjustment very small and slow to take effect.    Unclear how slow.

Another experiment is beginning now.   I've just added the multi-threaded miner to betanet as a guesser.  It's been finding blocks in 6-7 secs (guessing time).    So let's leave it running 24 hours or so and see how much that goes up.

**Comment by dan-da** [comment]
starting / reference point.   betanet.

block 781, difficulty: 22740569.  just under 4 secs guessing.

```
2024-11-26T17:03:55.824386129Z  INFO ThreadId(348) neptune_cash::mine_loop: Mining on block with 1 outputs and difficulty 22740569. Attempting to find block with height
 781 with digest less than target: 06936175366339482805,17550713868693256239,03582557654847030526,04389545200138075861,00000000811182168283                             
2024-11-26T17:03:59.524534119Z  INFO ThreadId(348) neptune_cash::mine_loop: Found valid block with nonce: (11959747264006501508,01836224421181250473,0813744012906203170
4,17338253617665495739,14511982029530850171).                                                                                                                           
2024-11-26T17:03:59.524616975Z  INFO ThreadId(348) neptune_cash::mine_loop: Newly mined block details:                                                                  
              Height: 781   
              Time  : 2024-11-26T18:03:59.521+01:00 (1732640639521)                                                                                                     
        Digest (Hex): 110dffa04063bf97d51f9a8701689adab30a7120f0efccfdeea7623b7c458ea2792ef142a4000000                                          
        Digest (Raw): 10934567549508979985,15752017002461011925,18288256001460210355,11713376080711821294,00000000705497738873    
Difficulty threshold: 06936175366339482805,17550713868693256239,03582557654847030526,04389545200138075861,00000000811182168283
          Difficulty: 22740569                                                                                                                                          
          #inputs   : 0                                                             
          #outputs  : 1
```

**Comment by dan-da** [comment]
subsequent block difficulties:
782: 23832426
783: 24846497
784: 25980008
785: 27215117
786: 28435949
787: 29759671

block 787 took about 40 secs.

so clearly it is adjusting, and much faster than in my test from genesis block with a single node composing and mining.

I think I will return this node to single-threaded mining now because otherwise the difficulty can rise so high that single-threaded nodes will have difficulty finding blocks on betanet.

**Comment by Sword-Smith** [comment]
According to many test runs, both unit tests and alpha/beta net, as long as the composing time is less than the target block interval, the difficulty adjusts such that the block time tracks the target interval.
